File: frisbee.cpp
#include <fstream> 
#include <iostream> 
#include <math.h>
#include <string>
#include <cstdlib>
using namespace std;

class Action
{
public:
  double x0;
  double y0;
  double vx0;
  double vy0;
  double g;
  double rho;
  double vx,vy,x,y;
  double m,area;
  double CLO,CLA,CDO,CDA;
  double alphaO;
  double alpha;
  double cl,cd,pi;
  double dt;
  ofstream out;
  void print();
  void iterate();
  Action(); //constructor
  ~Action();//Destructor
};

int main(void)
{
  Action A;
  
  A.print();
  while(A.y > 0)
    {
      A.iterate();
      A.print();
    };
    
}


Action::Action(){

  ifstream inF;
  inF.open("param.dat");
  if(inF.is_open())
    {
      inF >> alpha;
      inF >> x0;
      inF >> vx0;
      inF >> y0;
      inF >> vy0;
      inF >> dt;
    }
  else
    {cout << "Missing Input File. Exiting..." << endl; exit(1);}

  //Calculate other params
  g = -9.81;//gravity in m/s
  rho = 1.23;//air density in kg/m^3
  area = 0.0568;//area of average frisbee
  CLO = 0.1; // Lift Coefficient, alpha=0
  CLA = 1.4; // Lift Coefficient at Alpha
  CDO = 0.08; //Drag Coeff at alpha=0
  CDA = 2.72; //Drag Coeff at alpha
  alphaO = -4.;//characterizes angular dependence of 
               //drag due to frisbee shape
  m = .175;//average frisbee mass in kg
  pi = 3.1415926;

  vx = vx0;
  x = x0;
  y = y0;
  vy = vy0;

  //lift coefficient
  cl = CLO + CLA*alpha*(pi/180);

  //drag coefficient
  cd = CDO + CDA*pow((alpha-alphaO)*(pi/180),2);

  out.open("data.dat");
}

Action::~Action(){
  out.close();
}

void Action::print(){  
  out << x << " " << y << " " << vx << " " << vy << endl;
}

void Action::iterate(){
  double deltavx, deltavy;
  
  deltavx = -(rho*pow(vx,2)*area*cd*dt);
  deltavy = (((rho*pow(vx,2)*area*cl)/(2*m))+g)*dt;

  vx = vx + deltavx;
  vy = vy + deltavy;
  x = x + vx*dt;
  y = y + vy*dt;
}

File: frisbee.rb
class Action
  attr_accessor :x0, :y0, :vx0, :vy0, :g, :rho, :vx, :vy, :x, :y, :m, :area, :clo, :cla, :cdo, :cda, :alpha0, :alpha, :cl, :cd, :dt

  def initialize
    begin
      f = File.open("param.dat")
      line = f.readline
      self.alpha, self.x0, self.vx0, self.y0, self.vy0, self.dt = line.split.map(&:to_f)

      @out = File.open("data.rb.dat","w")
    rescue => e
      puts e
      exit
    end

    self.g = -9.81
    self.rho = 1.23
    self.area = 0.0568
    self.clo = 0.1
    self.cla = 1.4
    self.cdo = 0.08
    self.cda = 2.72
    self.alpha0 = -4.0

    self.m = 0.175

    self.vx = vx0
    self.x = x0
    self.y = y0
    self.vy = vy0

    self.cl = clo + cla*alpha*(Math::PI/180)

    self.cd = cdo + cda*(((alpha-alpha0)*(Math::PI/180))**2)

    @data = ""
  end

  #TODO: RENAME THIS!
  def print
    values = "#{x} #{y} #{vx} #{vy}"
    @data << values + "\n"
  end

  def run
    print
    while y > 0
      iterate
      print
    end
    @out << @data
  end

  private
    def iterate
      deltavx = -(rho*(vx**2)*area*cd*dt)
      deltavy = ((rho*(vx**2)*area*cl)/(2*m)+g)*dt

      self.vx = vx + deltavx
      self.vy = vy + deltavy
      self.x = x + vx*dt
      self.y = y + vy*dt
    end
end

a = Action.new

a.run
